### ISSUE TITLE
Reimplement Command Line API

### ISSUE THREAD

**Issue by SebastianZ** [firebug/firebug.next#361]
The [Command Line API](http://getfirebug.com/wiki/index.php/Command_Line_API) should be reimplemented.

This issue is meant to be a meta-issue for issues implementing parts of the API.

Commands, related issues and notes:
- [ ] `help`: already implemented, though refers to an MDN page
- [x] `$(selector)`: already implemented
- [x] `$$(selector)`: already implemented
- [ ] `$x(xpath[, contextNode[, resultType]])`: already implemented, though Firebug 2 offers advanced options, #365
- [x] `$0`: already implemented
- [ ] `$1`: not implemented yet
- [x] `$_`: already implemented, [bug 792063](https://bugzilla.mozilla.org/show_bug.cgi?id=792063)
- [ ] `$p`: not implemented yet, #230
- [ ] `$n(index)`: not implemented yet
- [ ] `dir(object)`: #244
- [ ] `dirxml(node)`: #244
- [x] `cd(window)`: already implemented
- [x] `clear()`: already implemented
- [x] `copy(object)`: already implemented, works differently than in Firebug 2
- [ ] `inspect(object[, panelName])`: already implemented, doesn't support panel name (#378)
- [x] `keys(object)`: already implemented
- [x] `values(object)`: already implemented
- [ ] `include(url[, alias]) / include(alias)`: #305 
- [ ] `debug(fn)`: not implemented yet #390 
- [ ] `undebug(fn)`: not implemented yet #390
- [ ] `monitor(fn)`: #366
- [ ] `unmonitor(fn)`: #366
- [ ] `monitorEvents(object[, types])`: #363
- [ ] `unmonitorEvents(object[, types])`: #363
- [ ] `profile([title])`: not implemented yet
- [ ] `profileEnd()`: not implemented yet
- [ ] `table(data[, columns])`: not implemented yet
- [x] `traceCalls(fn)`: not implemented yet (WONTFIX)
- [x] `untraceCalls(fn)`: not implemented yet (WONTFIX)
- [x] `traceAll()`: not implemented yet (WONTFIX)
- [x] `untraceAll()`: not implemented yet (WONTFIX)
- [ ] `getEventListeners()`: #364

Sebastian

**Comment by fflorent** [comment]
Thanks for this list Sebastian :)

Florent

**Comment by fflorent** [comment]
We will now be able to reimplement these commands when this patch will be checked-in (should normally be included in the upcoming nightly): 
https://bugzilla.mozilla.org/show_bug.cgi?id=1143497

Florent

**Comment by aliuk2012** [comment]
Nice work @fflorent well done.

**Comment by janodvarko** [comment]
> We will now be able to reimplement these commands when this patch will be checked-in
>  (should normally be included in the upcoming nightly): 

@fflorent  Great news!
Honza

**Comment by fflorent** [comment]
Merged : https://bugzilla.mozilla.org/show_bug.cgi?id=1143497

Also see : https://bugzilla.mozilla.org/show_bug.cgi?id=1151610 (Manage the case where two extensions fight over the same command)

Florent

**Comment by canuckistani** [comment]
The current state of the console object in nightly is:

* assert
* clear
* count
* debug
* dir
* dirxml
* error
* exception
* group
* groupCollapsed
* groupEnd
* info
* log
* markTimeline
* profile
* profileEnd
* table
* time
* timeEnd
* timeStamp
* timeline
* timelineEnd
* trace
* warn

In particular, we do have table, profile and profileEnd.

Some other notes:
* help() - see [bug 1161214](https://bugzilla.mozilla.org/show_bug.cgi?id=1161214)
* inspect() works, but probably slightly differently then FB users might expect.
* monitor / monitorEvents  / traceCalls / debug / traceAll - these are interesting but we've no plans currently that I know of to implement.
* getEventListeners() is pretty interesting.

Some other related bugs:
* [Use $variable in console](https://bugzilla.mozilla.org/show_bug.cgi?id=1154363)
* [Set $1 ... $n to recent expression results](https://bugzilla.mozilla.org/show_bug.cgi?id=1154359) - this is an interesting one because we've talked about mapping these to recent expression results instead of recent DOM nodes, following WebKit nightly. Still not clear what we're going to do but it's blocked on some UX improvements, eg if we magically assign something to a temp var, we should visually indicate what's been done.

**Comment by fflorent** [comment]
> In particular, we do have table, profile and profileEnd.

Right, you have console.table(), console.profile() and console.profileEnd(). And Firebug implement an alias of these command without having to prefix with `console.`. I think it is, at least, interesting to implement it for `profile[End]`. What do you think, @canuckistani ?

Florent

**Comment by fflorent** [comment]
@janodvarko, @SebastianZ The `traceCalls` command is a bit undocumented and I don't see what was the purpose of it or the added value compared to the `monitor()` command. Any idea? Or should we consider dropping it?

Florent

**Comment by SebastianZ** [comment]
@simonlindholm  explained what `traceCalls()` does in [issue #4315](https://github.com/firebug/firebug/issues/4315#issuecomment-103460179). Though I assume there is no additional value to `monitor()`, because the calls logged there already include the stack trace.

Sebastian

**Comment by fflorent** [comment]
Great, so I tick the `(un)traceCalls` command and mark it as WONTFIX. Thanks Sebastian!

Florent

**Comment by simonlindholm** [comment]
> @simonlindholm explained what traceCalls() does in issue #4315. Though I assume there is no additional value to monitor(), because the calls logged there already include the stack trace.

It's not the same - monitor(f) shows a stack trace when f is called, traceCalls(f) shows a tree-like structure of all execution (functions called) starting from f, but not including the stack trace of f itself. The tree is collapsed to avoid being unwieldy, quite comparably with the HTML panel. That said, I agree with scrapping traceCalls and traceAll; they are better implemented as UI in the debugger or profiler.

**Comment by fflorent** [comment]
Oh, now I see! There may be some use cases where this command can be interesting after all. But yeah, the profiler does the same job, as you said, so WONTFIXing sounds still reasonable.

Florent

**Comment by janodvarko** [comment]
All missing command line APIs should be reported in Bugzila and implemented on the platform side.

Honza

**Comment by SebastianZ** [comment]
I've [updated the list](https://github.com/firebug/firebug.next/issues/361#issue-62545363) and linked the related platform reports as far as they already exist.

Sebastian
